### PR TITLE
Option to filter short pathogen reference contigs

### PR DESCRIPTION
**Pull request by mwalker174**
In PathSeqBuildReferenceTaxonomy, adds option to exclude non-virus contigs below a specified length. This helps filter out a lot of false hits to unplaced contigs that were assembled from contaminant (usually human) sequences.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java [hunk 1]
```diff
@@ -212,20 +213,34 @@ private static void updateAccessionTaxonProperties(final String accession, final
...
+    protected static Map<String, String> buildAccessionToTaxIdMap(final Map<String, PSPathogenReferenceTaxonProperties> taxIdToProperties,
+                                                                  final PSTree tree,
+                                                                  final int minNonVirusContigLength) {
         final Map<String, String> accessionToTaxId = new HashMap<>();
         for (final String taxId : taxIdToProperties.keySet()) {
+            final boolean isVirus = tree.getPathOf(taxId).contains(VIRUS_ID);
             final PSPathogenReferenceTaxonProperties taxonProperties = taxIdToProperties.get(taxId);
-            for (String name : taxonProperties.accessions) {
-                accessionToTaxId.put(name, taxId);
+            for (String name : taxonProperties.getAccessions()) {
```
Comment (SHuang-Broad) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java]: `final String`

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java [hunk 2]
```diff
@@ -212,20 +213,34 @@ private static void updateAccessionTaxonProperties(final String accession, final
         } else {
             taxonProperties = new PSPathogenReferenceTaxonProperties();
```
Comment (SHuang-Broad) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java]: Several other places in this file could also be handled like this.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSPathogenReferenceTaxonProperties.java [hunk 3]
```diff
@@ -11,10 +14,34 @@
     public String name = null;
     public String rank = null;
     public String parentTaxId = null;
-    public long length = 0;
-    public final Set<String> accessions;
+    private long length = 0;
```
Comment (SHuang-Broad) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSPathogenReferenceTaxonProperties.java]: Not a comment on the `length` and `accessions` fields, but on the name and rank fields: is it possible to make them immutable  (final). I looked at where they are written, and it seems like they are written only when the default ctor of this class is used and the `name` and `rank` are immediately put after that. What about having a ctor that takes this two arguments?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java [hunk 4]
```diff
@@ -212,20 +213,34 @@ private static void updateAccessionTaxonProperties(final String accession, final
...
     }
 
     /**
+     * Removes nodes not in the tree from the tax_id-to-properties map
+     */
+    static void removeUnusedTaxIds(final Map<String, PSPathogenReferenceTaxonProperties> taxIdToProperties,
+                                   final PSTree tree) {
+        final Set<String> unusedNodes = new HashSet<>(taxIdToProperties.keySet());
+        unusedNodes.removeAll(tree.getNodeIDs());
+        taxIdToProperties.keySet().removeAll(unusedNodes);
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java]: Isn't this the same as `taxIdToProperties.keySet().retainAll(tree.getNodeIDs());`?

File: src/test/resources/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildReferenceTaxonomy/test.dict [hunk 5]
```diff
@@ -13,4 +13,4 @@
 @SQ	SN:ref|ACC_3c	LN:99	M5:b091f1dc6e3c6ef4391fabc5d8149f05	UR:file:/Users/markw/IdeaProjects/gatk/src/test/resources/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildReferenceTaxonomy/test.fasta
```
Comment (mwalker174) [inline src/test/resources/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBuildReferenceTaxonomy/test.dict]: This is a fair point but it's inconsequential (maybe the UR's are used in production?), plus it seems like all dict files in the repo have the same issue.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java (hunk 2 above)
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java]: Great. Did this here and a couple of other places where it was easy.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java (hunk 4 above)
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java]: Yep, thanks!

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java [hunk 6]
```diff
@@ -212,20 +213,34 @@ private static void updateAccessionTaxonProperties(final String accession, final
...
-        taxonProperties.accessions.add(name);
-        taxonProperties.length += length;
+        taxonProperties.addAccession(name, length);
         taxIdToProperties.put(accession, taxonProperties);
     }
 
     /**
+     * Removes nodes not in the tree from the tax_id-to-properties map
+     */
+    static void removeUnusedTaxIds(final Map<String, PSPathogenReferenceTaxonProperties> taxIdToProperties,
```
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java]: Test added and using @pshapiro4broad's suggestion below.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java (hunk 1 above)
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSPathogenReferenceTaxonProperties.java (hunk 3 above)
Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSPathogenReferenceTaxonProperties.java]: It appears that way but I sometimes set them directly, i.e. taxonProperties.name = "blah"

Comment (mwalker174) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSPathogenReferenceTaxonProperties.java]: Isn't quite possible we usually read in the taxonomic ID long before we get its name, in parseNamesFile(). I added get/set methods for PSPathogenReferenceTaxonProperties.

**Pull request merged**